components/example-notebook-servers/README.md

```diff
@@ -0,0 +1,139 @@
+
+#### Background
+
+It is often considered best practice to have only a single process run inside
+a container. A problem that can occur when using a process supervisor inside a
+container, is that the process supervisor will restart failed services and thus
+the container will never die. This is where the
+[S6-overlay](https://github.com/just-containers/s6-overlay) differs from other
+init systems, such as the popular [Tini](https://github.com/krallin/tini).
+While [Tini](https://github.com/krallin/tini) is created to handle a single
+process running in a container, the
+[S6-overlay](https://github.com/just-containers/s6-overlay) is built to manage
+multiple processes and allows the creator of the image to determine which
+processes should restart and which should cause the container to exit.
+
+#### Custom startup scripts
+
+Scripts that need to run during the startup of the Notebook Server container
+can be placed in the directory `/etc/cont-init.d/`. An example of such a script
+can be found [here](POINT TO RSTUDIO INIT SCRIPT). In this script, the
+`with-contenv` helper is used so that the container environment variables are
+available to the script. This script then exports the container environment
+variables to the R site environment so that they are available within RStudio.
+Without these environment variable, `kubectl` would not work in RStudio as
+`kubectl` relies on these environment variables to get the KUBECONFIG for the
+service account of the Notebook Server Pod. Scripts in the `/etc/cont-init.d/`
+directory are executed in order.
+
+#### Custom service scripts
+
+The [S6-overlay](https://github.com/just-containers/s6-overlay) manages the
+services that are started by the scripts present in the `/etc/services.d/`
+directory. Each service should have it's own folder in `/etc/services.d/` and
+contain a script named `run` and can optionally contain a finishing script
+called `finish`. For example, the provided Jupyter example images contain the
+file `/etc/services.d/jupyterlab/run` (LINK TO FILE) that the
+[S6-overlay](https://github.com/just-containers/s6-overlay) uses to start
+[JupyterLab](https://github.com/jupyterlab/jupyterlab). For more information on
+creating `run` and `finish` scripts for services please see the
+[S6-overlay documentation](https://github.com/just-containers/s6-overlay#writing-a-service-script).
+
+**Important**
+The example Notebook Server images that are provided here execute the
+[S6-overlay](https://github.com/just-containers/s6-overlay) as user `$NB_USER`
+and not as `root`. This is done so that none of the processes in the container
+are executed as `root` user. As such, any files or scripts related to the
+[S6-overlay](https://github.com/just-containers/s6-overlay) must be owner by
+user `$NB_USER`. To copy startup or service scripts into the container image
+you can use `--chown` with the `COPY` command in the Dockerfile
+(LINK TO A PLACE WHERE THIS IS USED). As the
```

Comment:
Missing URL (I know you know - I am just adding a review comment to make sure we don't forget about it)